lib/list.rb

```diff
@@ -2,38 +2,108 @@ class List
   include Enumerable
 
   def initialize(options={})
+    @options = options
+    @size = 0
   end
 
   # add object to list based on list type
   def add(object)
-    true
+    self.size += 1
+    if maintain_sort_order?
+      sorted_insert(object)
+    else
+      append(object)
+    end
   end
 
   # enumerable mixin method
   def each
+    if block_given?
+      current_node = head_node
+      while current_node
+        yield current_node.object
+        current_node = current_node.next
+      end
+    end
+    self
   end
 
   # returns all elements in list
   def elements
+    each.to_a
   end
 
   # returns true if the list is empty, false otherwise
   def empty?
-    true
+    !head_node
   end
 
   # returns length of list
   def length
-    0
+    size
   end
 
   # returns true if object is a member of the list, false otherwise
   def member?(object)
+    each { |obj| return true if object == obj }
+    false
   end
 
   # returns and removes the first element of the list; or nil if the list is empty
   def pop
-    nil
+    return nil if empty?
+    popped_node = head_node
+    self.head_node = head_node.next
+    self.size -= 1
+    popped_node.object
   end
 
-end
\ No newline at end of file
+  private
+
+  attr_reader :options
+  attr_accessor :head_node, :size
+
+  def maintain_sort_order?
+    sort_option = options[:sorted]
+    raise ArgumentError(sort_option) unless [ 0, 1, nil ].member?(sort_option)
```

Comment:
I add this check because the tests imply that the interface with the `:sorted` option is `0` for standard list, and `1` for a sorted list. This is not a very ruby way of doing things. In ruby, (just about) everything is an object and every object has a truthiness value. The keyword `false` and `nil` are the only falsey values. So both `0` and `1` are truthy values.

Therefore this interface is likely to confuse your average rubyist (especially if they are unfamiliar with languages where `0` is falsey).

However there is something to be said for stability of public interfaces. Perhaps many clients are already using this code, and we can't break backwards compatibility. In this PR, I decided to leave the interface as-is, but put in some tight error checking to hopefully save later developers some headache.

I will show my thoughts for improving the interface in a seperate PR.